fat_error_decrypt.go

```diff
@@ -0,0 +1,201 @@
+	// path.
+	Payloads [][]byte
+}
+
+// OnionFatErrorDecrypter is a struct that's used to decrypt fat onion errors in
+// response to failed HTLC routing attempts according to BOLT#4.
+type OnionFatErrorDecrypter struct {
+	fatErrorBase
+
+	circuit *Circuit
+}
+
+// NewOnionFatErrorDecrypter creates new instance of a fat error onion
+// decrypter.
+func NewOnionFatErrorDecrypter(circuit *Circuit,
+	structure *FatErrorStructure) *OnionFatErrorDecrypter {
+
+	return &OnionFatErrorDecrypter{
+		fatErrorBase: newFatErrorBase(structure),
+		circuit:      circuit,
+	}
+}
+
+// DecryptError attempts to decrypt the passed encrypted error response. The
+// onion failure is encrypted in backward manner, starting from the node where
+// error have occurred. As a result, in order to decrypt the error we need get
+// all shared secret and apply decryption in the reverse order. A structure is
+// returned that contains the decrypted error message and information on the
+// sender.
+func (o *OnionFatErrorDecrypter) DecryptError(encryptedData []byte) (
+	*DecryptedFatError, error) {
+
+	// Ensure the error message length is enough to contain the payloads and
+	// hmacs blocks. Otherwise blame the first hop.
+	if len(encryptedData) < minOnionErrorLength+o.hmacsAndPayloadsLen {
+		return &DecryptedFatError{
+			DecryptedError: DecryptedError{
+				SenderIdx: 1,
+				Sender:    o.circuit.PaymentPath[0],
+			},
+		}, nil
+	}
+
+	sharedSecrets, err := generateSharedSecrets(
+		o.circuit.PaymentPath,
+		o.circuit.SessionKey,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error generating shared secret: %w", err)
+	}
+
+	var (
+		sender      int
+		msg         []byte
+		dummySecret Hash256
+	)
+	copy(dummySecret[:], bytes.Repeat([]byte{1}, 32))
+
+	// We'll iterate a constant amount of hops to ensure that we don't give
+	// away an timing information pertaining to the position in the route
+	// that the error emanated from.
+	hopPayloads := make([][]byte, 0)
+	for i := 0; i < o.maxHops; i++ {
+		var sharedSecret Hash256
+
+		// If we've already found the sender, then we'll use our dummy
+		// secret to continue decryption attempts to fill out the rest
+		// of the loop. Otherwise, we'll use the next shared secret in
+		// line.
+		if sender != 0 || i > len(sharedSecrets)-1 {
+			sharedSecret = dummySecret
+		} else {
+			sharedSecret = sharedSecrets[i]
+		}
+
+		// With the shared secret, we'll now strip off a layer of
+		// encryption from the encrypted error payload.
+		encryptedData = onionEncrypt(&sharedSecret, encryptedData)
+
+		message, payloads, hmacs := o.getMsgComponents(encryptedData)
+
+		expectedHmac := o.calculateHmac(
+			sharedSecret, i, message, payloads, hmacs,
+		)
+		actualHmac := hmacs[i*sha256.Size : (i+1)*sha256.Size]
+
+		// If the hmac does not match up, exit with a nil message.
+		if !bytes.Equal(actualHmac, expectedHmac) && sender == 0 {
+			sender = i + 1
+			msg = nil
+		}
```

Comment:
Should we break or error here? Comment says that we exit, but we'll continue on to extract the payload.

Reply:
It looks a bit strange indeed, but the idea (copied from the legacy error decryption) is to decode in constant time. It seems a bit too much because there might very well be other loops in lnd over hops that aren't constant time. There is a comment right above the for loop explaining it.

Reply:
Agree that it's easy to miss the rationale above the for loop while reading the code here. Perhaps add a comment here as well.

Reply:
Added sentence to repeat that we first finish the constant number of iterations.